src/main/kotlin/org/rust/lang/core/resolve2/FacadeResolve.kt

```diff
@@ -52,30 +54,34 @@ fun processItemDeclarationsUsingModInfo(
     ipm: ItemProcessingMode
 ): Boolean {
     val (_, defMap, modData) = info
+    val elements = hashMapOf<RsNamedElement, Pair<EnumSet<Namespace>, VisibilityFilter>>()
     for ((name, perNs) in modData.visibleItems.entriesWithNames(processor.names)) {
-        // We need a `Set` here because item could belong to multiple namespaces (e.g. unit struct)
-        // Also we need to distinguish unit struct and e.g. mod and function with same name in one module
-        val elements = hashSetOf<RsNamedElement>()
         for ((visItems, namespace) in perNs.getVisItemsByNamespace()) {
             if (namespace !in ns) continue
             for (visItem in visItems) {
                 if (ipm == WITHOUT_PRIVATE_IMPORTS && visItem.visibility == Visibility.Invisible) continue
                 if (namespace == Namespace.Types && visItem.visibility.isInvisible && name in defMap.externPrelude) continue
                 val visibilityFilter = visItem.visibility.createFilter()
                 for (element in visItem.toPsi(info, namespace)) {
-                    if (!elements.add(element)) continue
-                    if (processor.process(name, element, visibilityFilter)) return true
```

Comment:
Why can't we use

```kotlin
if (processor.process(name, element, element.namespaces, visibilityFilter)) return true
```

Reply:
This test would not pass then:

```rust
mod a {
    pub struct S;
}
use a::*;
struct S {}
     //X
trait Trait<T> {}
impl Trait<S> for S {}
         //^
```

We should use actual namespaces in which the item is available in the scope, not scopes in which the item is defined